docs/en/03_Upgrading/index.md

```diff
@@ -37,123 +89,1015 @@ and also checking the `composer.json` files in [recipe-core](https://github.com/
+
+The best way of achieving this is to use the [`::class` PHP magic class constant](http://php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.class) which will return the fully qualified name of a class.
+
+Our example could be update to:
+```php
+<?php
+namespace Acme\Web\Products;
+
+use SilverStripe\ORM\DataObject;
+use SilverStripe\Assets\Image;
+use SilverStripe\Blog\Model\BlogPost;
+use Acme\Web\ProductService;
+
+class ExplosiveTennisBall extends DataObject
+{
+    private static $has_one = [
+        'Thumbnail' => Image::class
+    ];
+
+    private static $has_many = [
+        'Tags' => BlogPost::class
+    ];
+
+
+    public function getShippingCost()
+    {
+        return Injector::inst(ProductService::class)->calculateCost($this);
+    }
+}
+```
+
+Alternatively, you can spell out the full qualified name of each class in a string. For example, `'Image'` would become `'SilverStripe\\Assets\\Image'`. Note the use of the _double backslash_ — this is necessary because the backslash is an escape character.
+
+#### Update references to classes in your YML config
+
+YAML configuration files can reference SilverStripe classes. Those references also need to use the fully qualified name of each class.
+
+Take for example the following SilverStripe 3 YAML configuration file.
+
+```yaml
+Injector:
+  ProductService:
+    properties:
+      RoadRunnerSpeed: 99999999
+      CoyoteSpeed: 1
+
+BlogPost:
+  extensions:
+    - HasOneExplosiveTennisBallExtension
+
+Email:
+  admin_email: no-reply@example.com
+```
+
+In SilverStripe 4, this will become:
+```yaml
+SilverStripe\Core\Injector\Injector:
+  Acme\Web\ProductService:
+    properties:
+      RoadRunnerSpeed: 99999999
+      CoyoteSpeed: 1
+
+SilverStripe\Blog\Model\BlogPost:
+  extensions:
+    - Acme\Web\Extensions\HasOneExplosiveTennisBallExtension
+
+SilverStripe\Control\Email\Email:
+  admin_email: no-reply@example.com
+```
+
+#### Update references to classes in your language files
+
+Translation keys are normally tied to classes. If you override SilverStripe's default translation or if you are localising your own project, you'll need to update those references to use the fully qualified name of each class.
+
+For example, let's say you had the following translation file in `mysite/lang/eng.yml`.
+```yaml
+en:
+  Member:
+    SINGULARNAME: Coyote
+  RoadRunner:
+    SALUTATION: Beep Beep
+```
+
+In SilverStripe 4, it would become:
+```yaml
+en:
+  SilverStripe\Security\Member:
+    SINGULARNAME: Coyote
+  Acme\Web\RoadRunner:
+    SALUTATION: Beep Beep
+```
+
+### Finalising namespace updates
+Whatever you upgraded your code automatically or manually you'll need to perform the two following steps manually.
+
+DataObject tables will default to use a namespaced name. For example, if you have class under `Acme\Web\Products\ExplosiveTennisBall` that extends `DataObject`, the matching table in your database will be called `Acme_Web_Products_ExplosiveTennisBall`.
```

Comment:
make it clear here than "tables" means "database tables".

Also, why doesn't the upgrader add BC table names to classes?

Reply:
> Whatever you upgraded

Whether you upgraded?

Reply:
I've added the _database_ table reference.


@dhensby Your comment about the upgrader not adding the  `private static $table_name` reference is relevant. From memory SS4.0 wouldn't warn you if you didn't specify  a `private static $table_name`, but SS4.1 does.

Might be worthwhile adding it to the upgrade command.

Reply:
Added an issue about the upgrader https://github.com/silverstripe/silverstripe-upgrader/issues/104